imap_processing/codice/codice_l1a.py

```diff
@@ -262,33 +274,73 @@ def get_esa_sweep_values(self) -> None:
+        """
+        self.compression_algorithm = constants.HI_COMPRESSION_ID_LOOKUP[self.view_id]
+
+        # Decompress the binary string
+        science_values_decompressed = decompress(
+            science_values, self.compression_algorithm
+        )
+
+        # Divide up the data by the number of priorities or species
+        chunk_size = len(science_values_decompressed) // self.num_counters
+        science_values_unpacked = [
+            science_values_decompressed[i : i + chunk_size]
+            for i in range(0, len(science_values_decompressed), chunk_size)
+        ]
+
+        # TODO: Determine how to properly divide up hi data. For now, just use
+        #       arrays for each counter
+        self.data = science_values_unpacked
+
+    def unpack_lo_science_data(self, science_values: str) -> None:
+        """
+        Decompress, unpack, and restructure CoDICE-Lo data arrays.
+
+        The science data within the packet is a compressed, binary string of
+        values. These data need to be divided up by species or priorities,
+        and re-arranged into 2D arrays representing energy and spin angle.
 
         Parameters
         ----------
         science_values : str
             A string of binary data representing the science values of the data.
         """
         self.compression_algorithm = constants.LO_COMPRESSION_ID_LOOKUP[self.view_id]
-        self.collapse_table_id = constants.LO_COLLAPSE_TABLE_ID_LOOKUP[self.view_id]
 
-        # TODO: Turn this back on after SIT-3
-        # For SIT-3, just create appropriate length data arrays of all ones
+        # Decompress the binary string
+        science_values_decompressed = decompress(
+            science_values, self.compression_algorithm
+        )
+
         # Divide up the data by the number of priorities or species
-        # science_values = packets[0].data["DATA"].raw_value
-        # num_bits = len(science_values)
-        # chunk_size = len(science_values) // self.num_counters
-        # self.data = [
-        #     science_values[i : i + chunk_size] for i in range(0, num_bits, chunk_size)
-        # ]
-        self.data = [["1"] * 128] * self.num_counters
+        chunk_size = len(science_values_decompressed) // self.num_counters
+        science_values_unpacked = [
+            science_values_decompressed[i : i + chunk_size]
+            for i in range(0, len(science_values_decompressed), chunk_size)
+        ]
+
+        # Further divide up the data by energy levels
+        # The result is a [12,128] array representing 12 spin angles and 128
+        # energy levels
+        self.data = []
+        for counter_data in science_values_unpacked:
+            data_array = [
+                counter_data[i : i + self.num_energy_steps]
+                for i in range(0, len(counter_data), self.num_energy_steps)
+            ]
+            self.data.append(data_array)  # type: ignore[arg-type]
```

Comment:
Are you able to allocate the `(12, 128)` array initially and fill that in?

I'd find this a bit easier to follow as a double for loop over `(i, j)` and grabbing the data from the proper location as you go.

```python
data = np.empty(12, 128, dtype=...)
for i in range(12):
    chunk = science_values_decompressed[i*chunk_size:(i+1)*chunk_size]
    for j in range(128):
        # Turn it into an integer value as well?
        data[i, j] = int(chunk[j*num_energy_steps : (j+1)*num_energy_steps], 2)
```

Reply:
I think this is not quite right because we first need to chunk of the data by the number of counters (a variable number) before re-arranging the data by number of spin sectors (12) and number of energy levels (128). But I see what you are saying and I will play around with this.

To be honest, I find your solution to be harder to read/understand :joy: but maybe we can find some middle ground.

Reply:
Also I don't believe the values need to be converted to integers here. They should already be converted to a list of integers from the decompression algorithm.

Reply:
I might not be following what is going on here then. I thought you were grabbing ranges out of a string to try and put groups of "101010" together appropriately.

I hear you on some of it being personal preference with how to write nested for loops, so no worries to keep it how it makes the most sense to you, mine are purely suggestions.

> I think this is not quite right because we first need to chunk of the data by the number of counters (a variable number) before re-arranging the data by number of spin sectors (12) and number of energy levels (128). But I see what you are saying and I will play around with this.

It might help to put in the input shape / an example of what you are getting, and what you are trying to turn it into as a comment above this. If the number of counters is variable, does that mean that you have a 3D array coming out? So a `data.reshape(-1, 12, 128)` would get you an (ncounters, 12, 128) shaped array output.

Reply:
> I might not be following what is going on here then. I thought you were grabbing ranges out of a string to try and put groups of "101010" together appropriately.

That was true recently, but with this PR I changed the decompression algorithm to spit out lists of (decompressed) integers instead, so that is now taken care of in the `decompress()` function call.

I took your latest suggestion as inspiration and came up with this:

```python
    def unpack_lo_science_data(self, science_values: str) -> None:
        """
        Decompress, unpack, and restructure CoDICE-Lo data arrays.

        The science data within the packet is a compressed, binary string of
        values. These data need to be divided up by species or priorities (or
        what I am calling "counters" as a general term), and re-arranged into
        2D arrays representing spin sector and energy.

        Parameters
        ----------
        science_values : str
            A string of binary data representing the science values of the data.
        """
        self.compression_algorithm = constants.LO_COMPRESSION_ID_LOOKUP[self.view_id]

        # Decompress the binary string into a list of integers
        science_values_decompressed = decompress(
            science_values, self.compression_algorithm
        )

        # This will hold the data to be later stored in the CDF. Each element
        # of the list will store the data for a particular counter
        self.data = []

        # Divide up the data by the number of counters
        chunk_size = len(science_values_decompressed) // self.num_counters
        science_values_per_counter = [
            science_values_decompressed[i : i + chunk_size]
            for i in range(0, len(science_values_decompressed), chunk_size)
        ]

        # Re-arrange the counter data to be a 2D array with a shape representing
        # [<num_spin_sectors>,<num_energy_steps]
        for counter_data in science_values_per_counter:
            reshaped_data = np.array(counter_data, dtype=np.int64).reshape(self.num_spin_sectors, self.num_energy_steps)

            # Add the re-arranged data to the list
            self.data.append(reshaped_data)
```

I also added some clarifying in-line comments and improved some variable names. Does this read well to you?

Reply:
Yes, I like that 👍 good idea!

I'm not sure about the unpacking order but I still think you might be able to make `self.data` an array instead of a list of arrays? 
```
np.array(science_values_decompressed, dtype=np.int64).reshape(self.num_counters, self.num_spin_sectors, self.num_energy_steps)
```

Do you want `int64` here or are the counts uint and smaller size? (Can be updated later and use Tim's dtype idea within the metadata too if that is helpful)

Reply:
My motivation for making `self.data` a list here is because when I go to create the CDF, I then iterate over the list to create separate data variables for each counter. Is it better to iterate over a list or over a dimension of a `np.array`?

Good call on the `int64`, I think I can use `uint` here instead. I had just used `int64` without thinking much about it.

Reply:
Ahh, I see what you're saying now that they are all separate "variables" in your output file.
> Is it better to iterate over a list or over a dimension of a np.array

It is almost always better/faster to go to numpy arrays if possible (here you'd be creating one instead of ncounter arrays). So writing out the counters in your future steps I'd do something like:
```python
for i in range(self.num_counters):
    counter_variable = self.data[i, ...]  # (12, 128)
```
which should be very similar to your current way with lists.

Reply:
This got me thinking, could this whole logic just be reduced to:

```python
self.data = np.array(science_values_decompressed, dtype=np.uint).reshape(self.num_counters, self.num_spin_sectors, self.num_energy_steps)
```

And avoid having to go over the for loops/chunking to restructure the data? I don't have a full understanding of how `np.reshape` is working under-the-hood, and if it is actually re-arranging the values how I would want it to.

Reply:
Thanks for the validation here. I am glad I spent an hour this morning accidentally re-implementing `numpy.reshape` from scratch 😅 What a fun exercise.

Reply:
> I am glad I spent an hour this morning accidentally re-implementing numpy.reshape from scratch 😅 What a fun exercise.

We've all been there! 😄  I don't think it was wasted time, because now I'm guessing you have a better appreciation of what is all going on under the hood and how the data is organized internally as well.